### PR TITLE
Update docstrings in `Field`

### PR DESCRIPTION
**Pull request by johnzl-777**
Follow the format proposed in #929 for spatial modulation options.

See here for preview: https://bloqade.quera.com/pr-preview/pr-930/reference/bloqade/builder/field/

### CODE REVIEW

File: src/bloqade/builder/field.py [hunk 1]
```diff
@@ -64,99 +130,292 @@ def location(
...
     ) -> "Location":
-        """Address a single atom (or multiple) atoms.
+        """Address a a subset of atoms as part of defining the spatial modulation component of a drive
+        with a unique scaling factor for each of them.
+
+        Args:
+            labels (Union[List[int], int]): A list of site indices or a single site index to target.
+            scales (Union[List[ScalarType], ScalarType, None]): A list of scaling factors for each site or a single scaling factor to target all sites.
+
+        Returns:
```
Comment (Roger-luo) [inline src/bloqade/builder/field.py:140]: This is a mix of RST (numpy style doc) and Markdown, we should stick with Markdown syntax.

File: src/bloqade/builder/field.py [hunk 2]
```diff
@@ -64,99 +130,292 @@ def location(
...
+            ```
+
+            And finally we can target a subset of sites with custom scaling coefficients:
+            ```python
+            custom_multi_location_scaled = program.location([0, 2], [0.5, 0.25])
+            ```
+        
+        ??? warning "Potential Pitfalls"
+
+            # Global vs. Local Detuning
```
Comment (Roger-luo) [inline src/bloqade/builder/field.py:200]: this might be too large after render? <h1> seems to be a pretty large thing to me.

Comment (johnzl-777) [reply src/bloqade/builder/field.py:200]: I've tested this locally and the rendering seems to be fine on the website. Inside admonitions the text in general seems to be smaller which makes this a reasonable looking headline. 

Check out the screenshot below:
Screenshot 2024-04-29 at 8 21 19 AM

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (johnzl-777) [reply src/bloqade/builder/field.py:140]: Could you give an example of a pure markdown format for the above? I thought this format was markdown compatible (I believe I'm following Google style here). RST style looks drastically different from this according to the following: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

File: src/bloqade/builder/field.py [hunk 3]
```diff
@@ -14,26 +14,92 @@ def uniform(self) -> "Uniform":
...
+            * [Nonequilibrium Dynamics of nearly Blockaded Rydberg Atoms](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-2-nonequilibrium-dynamics-blockade-radius/)
+            * [1D Z2 State Preparation](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-3-time-sweep/)
+            * [2D State Preparation](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-3-2d-ordered-state/)
+            * [Quantum Scar Dynamics](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-4-quantum-scar-dynamics/)
+            * [Solving the Maximal Independent Set Problem on defective King Graph](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-5-MIS-UDG/)
+            * [Lattice Gauge Theory Simulation](https://queracomputing.github.io/bloqade-python-examples/latest/examples/example-6-lattice-gauge-theory/)
+
+        ??? warning "Potential Pitfalls"
+
+            # Global vs. Local Detuning
```
Comment (weinbe58) [inline src/bloqade/builder/field.py:77]:
```suggestion
            # Global vs. Local Detuning on Hardware
```

Just make it clear from the title that this is hardware specific issue.

File: src/bloqade/builder/field.py [hunk 4]
```diff
@@ -64,99 +130,292 @@ def location(
...
-        as a list of labels and a list of scales. The scales are used to
-        multiply the waveform that is applied to the atom. You can also specify
-        a single label and scale to target a single atom.
+            [`uniform`][bloqade.builder.field.Detuning.uniform] and using `location` where all atoms are targeted with a scaling of $1$ or [`scale`][bloqade.builder.field.Field.scale]
+            with similar behavior in the construction of the Detuning field will lead to identical behavior in emulation but can have subtle differences on hardware as [`location`][bloqade.builder.field.Field.location]
+            and [`scale`][bloqade.builder.field.Field.scale] use a locally targeting laser while `uniform` uses a globally (all-atom) targeting laser.
+
+            # Number of Coefficients
+
+            The number of coefficients you provide should be equal to the number of atoms in your program. The first coefficient will target the first atom, the second coefficient will target the second atom, and so on.
```
Comment (weinbe58) [inline src/bloqade/builder/field.py:208]:
```suggestion
            The number of coefficients you provide should equal the number of sites in your program, even if some sites are not filled with an atom. The first coefficient will target the first atom, the second coefficient will target the second atom, and so on.
```

**Review by weinbe58 — APPROVED** [review]
some minor changes I saw.

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (Roger-luo) [reply src/bloqade/builder/field.py:140]: The google style is not in markdown? E.g Args needs to be specified as a title instead of indent?

Comment (Roger-luo) [reply src/bloqade/builder/field.py:140]: more specifically, the `Args:` syntax is from RST (see from your link). In markdown, this should be `### Args` (pick a title number here, I usually use the <h3> tag).

Comment (Roger-luo) [reply src/bloqade/builder/field.py:140]: here is a screenshot from vscode hint, since it's not markdown you lost the highlights

image

File: src/bloqade/builder/field.py (hunk 2 above)
Comment (Roger-luo) [reply src/bloqade/builder/field.py]: this is too large in vscode

image

**Comment by Roger-luo** [comment]
also I'm not sure which one is the correct syntax, either `!!!` or `???` but vscode is not recognizing `???` apparently. The github markdown uses `!!!` or something else. It'd be nice to make vscode render things consistently, but not a blocking requirement. Fixing the title size is a requirement, because that effects many docstring rendering.

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (johnzl-777) [reply src/bloqade/builder/field.py:140]: Ah I understand now. The problem is if I introduce the markdown `###` the rendering on MkDocs starts to fall apart:

Screenshot 2024-05-01 at 8 53 47 AM

versus what it looks like without the `###` 
Screenshot 2024-05-01 at 9 00 09 AM

File: src/bloqade/builder/field.py (hunk 2 above)
Comment (johnzl-777) [reply src/bloqade/builder/field.py]: I see, you're talking about the headers that are outside the admonition box. I agree it's cartoonishly large in VSCode but I just worry the titles are now too small on MkDocs.

If I use `##` or `###` instead of `#` the MkDocs stuff comes out as the following:
Screenshot 2024-05-01 at 9 09 07 AM

Instead of: 
Screenshot 2024-05-01 at 9 09 27 AM

Comment (Roger-luo) [reply src/bloqade/builder/field.py]: hmm, there seems to be a conflict here, but IMO vscode is doing the right convention of title size...

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (Roger-luo) [reply src/bloqade/builder/field.py:140]: maybe should be https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#definition-lists ?

**Comment by Roger-luo** [comment]
maybe we should check what are other projects using when combined with mkdocs-material markdown docstring.

**Comment by Roger-luo** [comment]
> Is that in reference to your previous comment where "Reachable Via" and "Next Possible Steps" render too largely in VSCode?

Yes. But VSCode is doing the correct thing. But better double check what gets lowered to by mkdocs, is it `<h1>` or something else.

**Comment by johnzl-777** [comment]
@Roger-luo I did an HTML/CSS inspect, oddly enough it gets rendered as `<h4>` by mkdocs

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (weinbe58) [reply src/bloqade/builder/field.py:140]: @Roger-luo The `???` isn't creating...I don't think VSCode highlights `Args` and `Returns` just try it out on a simple example.

**Comment by weinbe58** [comment]
@Roger-luo can we address these issues later? We're about to have a hackathon and I really doubt the participants will be using a lot of VSCode and even so I think its OK if we don't have perfect docs the website is more important.

File: src/bloqade/builder/field.py (hunk 1 above)
Comment (Roger-luo) [reply src/bloqade/builder/field.py:140]: Because it's not markdown, vscode only highlights markdown titles. Which is why I'm asking for changing it to title instead of RST here.

**Comment by Roger-luo** [comment]
I'm not sure why someone would spend that much time on websites instead of pop up hints from vscode. This is against the original design of builders - minimizing the need of going back forth and search a website. This is why I'm more emphasizing the vscode docstring. And if this is actually happening something must be wrong. And we should not spending time making website references renders better but to figure out why our user is coming back to the website.

**Comment by Roger-luo** [comment]
I'm more concerned that now as a real user in vscode I have to scroll all the way down to know that I can do next.

**Comment by weinbe58** [comment]
Talk to @plquera he is building out education platform, the vast majority of which is Jupyter notebooks as far as I understand

**Comment by Roger-luo** [comment]
@plquera can you chime in here? It's not clear to me what issue these changes are trying to address.

TL;DR, if you want to explain the concept they should be in a more highlighted position first instead of docstring. This deserves a section in the documentation instead of hiding in the reference.

And this makes me feeling we are teaching things in a non-idiomatic fashion. Thus we are solving an XY problem here. I believe it is important to be consistent and emphasize how one can build more complicated programs in a more serious fashion that is not just notebooks teaching toy examples. I don't think improving rendering of the references is the way to go in general. I have been saying we should have a quick start guide collecting common questions and explaining common concepts. This should have higher priority than the reference. Otherwise things are hidden in the docstring would make it too dense when someone relatively familiar with the physics want to code.

@weinbe58 sure if you can change the order, I'm temporarily fine with the markdown title of Arg and Return. But that's not what we should adopt without further looking.

**Comment by plquera** [comment]
Chiming in a little too late.

While I agree that visiting the website may not be what was envisaged in the original architected programming experienced, it is just a reality. When I see users trying to get into Bloqade, they need something a little more solid to hold their hands than  the pop-up hints. They need to consult slowly at what things mean and how to work with them. 

Besides that, it is true that notebooks are used more often than VSCode (although sometimes VSCode + notebooks work fine with pop-up hints). 

To decide the better UX design parameters, we need to run more experiments with more users. Having more folks from the scientific software team involved in the hackathons can help align and provide you with a first-hand view of the experience the students go through. 

For now, my suggestion is: let's update the docs. We should see them as complementary to the pop-up hints anyway, and John's contributions - probably not yet perfect - really improve the general content. 

Let's then also monitor how the participants of the ETH hackathon engage with Bloqade, hints, and the new documentation. I will try to take notes and bring for a debrief soon. Iteration is key.

**Comment by Roger-luo** [comment]
> While I agree that visiting the website may not be what was envisaged in the original architected programming experienced, it is just a reality. When I see users trying to get into Bloqade, they need something a little more solid to hold their hands than the pop-up hints. They need to consult slowly at what things mean and how to work with them.

I agree that "they need something a little more solid to hold their hands than the pop-up hints". This is what examples are for. And that's not what pop-up hints are for. The pop-up hints are just for the programming experience. The PR is trying to solve the problem in the wrong place. The sections with `???` are too dense with the hints. If users feel alienated by the concepts, we should explain the examples they are trying instead of asking them to search on the documentation website. This is +0 clicks (proposed) vs +5 clicks (this PR, open browser, type, press search, scroll, and click on the hidden info); this is not great UX. I believe we are now adding too many steps for a newcomer user.

Taking https://github.com/QuEraComputing/bloqade-python/pull/930/files#diff-45a36aa626cf04f249c77836985d1bb64fbe3e27e5cf0412e9c2c2b56d0faed1R23-R47 as an example, this information is hidden by default on the website, and the content seems only applicable for newcomers, why don't we just put these explanations in a quick start guide? I doubt this will be an improvement in UX. 

> Besides that, it is true that notebooks are used more often than VSCode (although sometimes VSCode + notebooks work fine with pop-up hints).

This is what I am concerned about, the indicates we have less serious users, and bloqade is currently mainly used for education. We are targeting different groups of users.

> For now, my suggestion is: let's update the docs. We should see them as complementary to the pop-up hints anyway, and John's contributions - probably not yet perfect - really improve the general content.

My job is to ensure this doesn't end up in a trash mountain; thus, I prefer a small, simple solution. If the content is too dense, it would scare people too. Maybe I'll just do this myself afterward to show you guys what I am thinking exactly.

---

Ok for @johnzl-777, given the time constraints you have (I really don't like this, I've approved too many workaround PRs this year). Currently, can you use the format we discussed above in the following order? But I reserve the option of moving the `???` sections to their page with more highlights from the homepage. And keep method docstrings simple.

- Args
- Kwargs
- Returns
- Next Possible Steps
- Reachable Via
- ???s

I'm good with this PR once these changes are made, and you can keep revising using this format.

**Comment by Roger-luo** [comment]
Apologies if someone thinks I meant the PR is trash by saying "trash mountain"; that is not my intention. I don't want things to get more complicated than they should be. Long docstrings are very hard to maintain within code because they are part of the code, you have to go through it in a code review. The explanation of abstraction itself should not belong here. I really meant the following:

The users who will see the references are from IDEs such as vscode by that assumption, they have already learned the basic concepts. The explanation in `???` is not useful to them. However, on the website, the `???` section is hidden, so an entry-level user cannot see it. This wastes the work done on the content (I like the content). I'd prefer this to be more highlighted on the front page, as I talked multiple times about a quick start guide on the homepage. Explain the abstraction there, that is, where one targets the entry-level user.

From Slack, I see the intention is to replicate Wolfram Mathematica from @plquera, I'm copying my reply for future reference

1. Wolfram docs are not designed for Python, as you can click on the functions and look for references within a notebook. this is not doable in Python. 
3. Wolfram is not a good software for developing more serious software work (unless you can find a serious software written using wolfram).

we should really look at how the Python community, especially the scientific community, is doing the docs. Spend the time on what brings more value.

**Comment by plquera** [comment]
I agree with much of this last comment from Roger.
We should strive for simpler pages with more simplified information.

The one things I disagree:
- Wolfram documentation works very well, independently of the function clicking. It is about what content is necessary and sufficient for users to easily get going, and reference around. 
- While Wolfram is not amazing for production software, it works perfectly well for research and development. I strongly advocate that quantum software should not be made for serious developers at this moment. The user experience should, in fact, target people that do not like to code. Making it easy and usable for the people innovating on math and algorithm development should help increase our user base in the short term. As tech matures, we can start aiming at professional developers. But professional developers are *not* the majority of today's users.

**Comment by Roger-luo** [comment]
While I'm not disagreeing your points about Wolfram, the conflict is really that python is not a language designed only for non-developers. Its tooling is strongly backed by developer tooling. This is the major difference. This is why I'm saying we should learn from other scientific packages such as pytorch. They have a lot of non-developer users and apparently people also love their documentation. We are not the first ones who think about this, but looking at other's solutions is better than being creative in the first place. We should learn the idiomatic approach first before being creative and improving it. 

(This PR is an example, there is a reason why people don't put an entire article into the reference like Wolfram).

**Comment by johnzl-777** [comment]
I see that this proposed method of documentation has caused a great deal of disagreement among the team. I accept some responsibility here for having jumped the gun and not gotten everybody on the same page. 

I think for the sake of moving forward that further discussion is certainly necessary and I am more than willing to close this PR and those with formats duplicating #929 for now. 

Legitimate points have been levied by both sides but for the sake of any future development of Bloqade documentation I do not want to make any contributions until an agreement is made.

**Comment by johnzl-777** [comment]
Not deleting the branches because the material could be reused in the future
